Mathlib/CategoryTheory/Limits/EpiMono.lean

```diff
@@ -0,0 +1,106 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Limits.Shapes.CommSq
+
+/-!
+# Relation between mono/epi and pullback/pushout squares
+
+In this file, monomorphisms and epimorphisms are characterized in terms
+of pullback and pushout squares. For example, we obtain `mono_iff_isPullback`
+which asserts that a morphism `f : X ⟶ Y` is a monomorphism iff the obvious
+square
+
+```
+X ⟶ X
+|   |
+v   v
+X ⟶ Y
+```
+
+is a pullback square.
+
+
+-/
+
+namespace CategoryTheory
+
+open Category Limits
+
+variable {C : Type*} [Category C]
+
+lemma mono_iff_fst_eq_snd {X Y : C} {f : X ⟶ Y} {c : PullbackCone f f} (hc : IsLimit c) :
+    Mono f ↔ c.fst = c.snd := by
+  constructor
+  · intro hf
+    simpa only [← cancel_mono f] using c.condition
+  · intro hf
+    constructor
+    intro Z g g' h
+    obtain ⟨φ, rfl, rfl⟩ := PullbackCone.IsLimit.lift' hc g g' h
+    rw [hf]
+
+lemma mono_iff_isIso_fst {X Y : C} {f : X ⟶ Y} {c : PullbackCone f f} (hc : IsLimit c) :
```

Comment:
Does it make sense to also state the version with `IsIso c.snd`? Does it help to prove a TFAE statement, and possibly deduce these lemmas as 1-line corollaries?

Reply:
In this particular case, it seems I would have to show the exact same number of implications in order to prove the TFAE, so that it would not be a shortcut.

I have added the `IsIso c.snd` variant, but in order to do that, I had to slightly refactor the flip of pullback cones and pushout cocones.